spec.md

```diff
@@ -96,7 +96,63 @@ and eventually have it accepted as a Unicode Technical Standard (UTS).
 
 ## Data Model
 
-> _The data-only representation of a message._
+A Message provides the representation of a single message.
+It takes one of two forms,
+either as a PatternMessage or a SelectMessage.
+In a multi-locale context,
+a single message's choice of representation as one or the other
+should not affect its representation in other languages or locales.
+
+```ts
+type Message = PatternMessage | SelectMessage
+```
+
+A PatternMessage contains a list of PatternElement values,
+some of which are directly defined literal values,
+while others are placeholders with formatted values that depend on additional data.
+
+```ts
+interface PatternMessage {
+  type: 'message'
+  value: PatternElement[]
+  meta?: Meta
+}
+```
+
+SelectMessage provides for the selection of one list of PatternElement values
+to use as the message's value when formatting,
+depending on the value of one or more Selector values.
+
+Each SelectCase is defined by a `key` of one or more string identifiers,
+and selection between them is made according to the corresponding Selector values.
+From this it follows that a valid SelectMessage must have at least as many `select` entries
+as its highest count of SelectCase `key` entries within its `cases`.
+The `fallback` value of a Selector is used in addition to its `value`
+when selecting one of the `cases` during formatting.
+It should match exactly one of the corresponding SelectCase `key` values.
+
+As a SelectMessage is not a valid PatternElement,
+it is not possible for it to be placed within the `value` of another message.
+
+```ts
+interface SelectMessage {
+  type: 'select'
+  select: Selector[]
+  cases: SelectCase[]
+  meta?: Meta
+}
+
+interface Selector {
+  value: PatternElement
+  fallback?: string
+}
+
+interface SelectCase {
+  key: string[]
+  value: PatternElement[]
```

Comment:
That would be a possible alternative, but what would be the benefit?

Reply:
Semantics.

"This thing here is a simple message, same as the other"

The same reason people `#define` or `typedef` things in C/C++, or use `strong` instead of `bold`
"And array of X" is the same, but is it "the same thing" as a simple message?
Or just an implementation detail?

Reply:
It's not _quite_ the same. Specifically, `SelectMessage` and `SelectCase` may both define a `meta` value that needs to be accounted for e.g. while formatting. This means that the computed `meta` value of a selected case is built from two different sources. The `meta` value of a `PatternMessage` is coming from a single source.

Therefore, keeping the two structures as separate interfaces makes sense, as they need to be handled differently.

Reply:
Then I would define a type (MessageBody?) that is used in both places.
Higher level of abstraction.
MessageBody tells us more than an array of `PatternElement`
And it shares behavior, parsing, implementation.

We can say: these are correct ways to represent and handle the `MessageBody` vs describe how to correctly handle two arrays of PatternElement in two levels.

We name things and we rise the level of abstraction.

> may both define a meta value that needs to be accounted for e.g. while formatting. This means that the computed meta value of a selected case is built from two different sources. The meta value of a PatternMessage is coming from a single source

This is a good example of underspecified piece of data that "slips" in the data model without anyone knowing what's inside.

All of this: _"may define"_, the fact that it is possible to have _"computed meta"_ (what does it mean?), that _"coming from a single source"_, that _"meta value that needs to be accounted for e.g. while formatting"_ (so it influence rendering? Why is not in `options`?).

All of these are undocumented assumptions that only exist in one's mind.
Then we approve this, and two months down the line when we open the "Meta" can of worms and find problems the reaction is "sorry, but you approved it in September"

We should not approve undocumented black boxes.

In fact I would propose we remove the Meta for now.
I 100% agree that we need "something like it", but can't just put it the spec completely under the radar and without knowing what that is.

Reply:
> Then I would define a type (MessageBody?) that is used in both places.
> Higher level of abstraction.
> MessageBody tells us more than an array of `PatternElement`
> And it shares behavior, parsing, implementation.
> 
> We can say: these are correct ways to represent and handle the `MessageBody` vs describe how to correctly handle two arrays of PatternElement in two levels.
> 
> We name things and we rise the level of abstraction.

Is your request then to add a
```ts
type MessageBody = PatternElement[]
```
and to use that to type the `value` properties of `PatternMessage` and `SelectCase`?

> In fact I would propose we remove the Meta for now.
> I 100% agree that we need "something like it", but can't just put it the spec completely under the radar and without knowing what that is.

I would like to quote a part of the original description of this PR:

> The definitions of PatternElement and Meta are left outside the scope of this PR.

In other words, this PR makes absolutely no assertions about the definition or use of metadata values, other than noting (as you put it) that we need "something like it". Are there specific nodes that you think should not be able to contain metadata?

Reply:
I also had concerns about leaving `Meta` undefined initially, but I think it is too much to require everything to be fully defined before including it here. We have a process that allows for `Meta` to be changed or removed after it is written, so if we're in agreement that we need something like it, I don't think there is any harm in including it as a placeholder.

Reply:
I think that if we agree we "need something like it" but we don't know what it looks like, then we should add an issue, so we don't forget, and put it in the spec later.

But to say "this part of the spec should be that way because of `Meta`, but I don't tell you what `Meta` is or does" feels wrong.
If it affects decisions on the current part of the spec then it kind of has to be there.
Otherwise it is _"just trust me"_

What worries me is that we push this through, and later we discover that the `Meta` as envisioned in one person's mind has limitations, does not do the right things, there are better ways, whatever.
And then the argument will be: we can't change this part of the spec, you agreed with it.

Just live it out.

This is the stated goal of this PR:
> As a first step in actually filling out the spec, here's my proposal for the Message data model.
> 
> This should be relatively uncontroversial, as we've previously agreed that messages may have top-level selectors. This data model is also as simple as possible, and does not e.g. allow for number values in selector keys. I'll later submit a PR for a selector algorithm that works with the proposed data model.

This looks controversial, and was not previously agreed.

So, leave it out?

Reply:
> Is your request then to add a
> `type MessageBody = PatternElement[]`
> and to use that to type the value properties of PatternMessage and SelectCase?

Yes.

Reply:
> this PR makes absolutely no assertions about the definition or use of metadata values

Not the PR directly, but your comment, when you say "can't do X because of `Meta`"
So it is an unspecified behavior of meta that affects decisions in this PR.

Reply:
`meta` values dropped, `MessageBody` added.